### PR TITLE
WIP: Travis CI

### PR DESCRIPTION
**Pull request by mpranj**
Travis builds for OS X, thanks @manuelm a lot!
I opened the PR is just to confirm how many things I broke on linux builds.

### CODE REVIEW

**Comment by mpranj** [comment]
jenkins build bindings please

**Comment by mpranj** [comment]
it appears to me that 5e2adad6165e2b80216482b4786014b1d2b80b18 fixed java bindings tests on OS X ... 
Here's a log from travis:
```
        Start  63: testmod_jni
 63/128 Test  #63: testmod_jni ........................   Passed    0.66 sec
```

File: .travis.yml [hunk 1]
```diff
@@ -0,0 +1,64 @@
+#
+# Turn off email notifications entirely
+#
+notifications:
+  email: false
```
Comment (markus2330) [inline .travis.yml]: Why not sending notifications? Jenkins does this quite well (only mails when the build breaks)

If only one e-mail address is allowed we could use some forward e-mail address, e.g. travis@libelektra.org

File: .travis.yml [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+  include:
+#    - services: docker
+#      env: OS_RELEASE=debian_unstable
+#      sudo: required
+#      script:
+#        - docker-compose -f scripts/docker/docker-compose.yml build $OS_RELEASE
+#        - docker-compose -f scripts/docker/docker-compose.yml run $OS_RELEASE /bin/bash /workspace/code/scripts/travis.sh
+    - os: osx
+
+osx_image: xcode7.3 # use latest OS X
```
Comment (markus2330) [inline .travis.yml]: Does everyone has the latest OS X? Are we interested in building for older OS X, too?

File: .travis.yml [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+      which python3 || brew install python3 &&
+      which lua || brew install lua &&
+      brew install gobject-introspection &&
+      brew install yajl &&
+      brew install augeas &&
+      brew cask install java;
+    fi
+#
+# Source is checked out in $TRAVIS_BUILD_DIR
+# Build directory will be TRAVIS_BUILD_DIR/../build
```
Comment (markus2330) [inline .travis.yml]: I would prefer if you assign some variables for this, maybe some ROOT_DIR, BUILD_DIR, SOURCE_DIR and so on.

File: scripts/docker/Dockerfile.debian_unstable [hunk 4]
```diff
@@ -0,0 +1,2 @@
+FROM manuelm/libelektra_build:debian_unstable
```
Comment (markus2330) [inline scripts/docker/Dockerfile.debian_unstable]: Can we copy this to ElektraInitiative/libelektra_build:debian_unstable?

File: src/bindings/jna/CMakeLists.txt [hunk 5]
```diff
@@ -27,6 +27,21 @@ if (Java_VERSION_MAJOR GREATER 0 AND Java_VERSION_MINOR GREATER 7)
 		NAMES jna
 		PATHS /usr/share/java
 		)
+	
+	if(NOT jna) # try fetching it, remember to update SHA1 if you update the jar
+		file(DOWNLOAD 
+			"https://maven.java.net/content/repositories/releases/net/java/dev/jna/jna/4.2.2/jna-4.2.2.jar"
```
Comment (markus2330) [inline src/bindings/jna/CMakeLists.txt]: Is this jar then bundled with elektra.jar? Or how does the installed version work then?

File: src/bindings/jna/elektra/Elektra.java [hunk 6]
```diff
@@ -11,7 +11,7 @@
 	 * Overall native binding to the elektra library
 	 */
 	Elektra INSTANCE = (Elektra)
-		Native.loadLibrary(("libelektra.so"), Elektra.class);
+		Native.loadLibrary(("elektra"), Elektra.class);
```
Comment (markus2330) [inline src/bindings/jna/elektra/Elektra.java]: Good fix, seems to work with Linux, too.

**Comment by markus2330** [comment]
Nice work by both of you!

But the travis build itself fails ;( Should we exclude the py/lua bindings or can someone fix the LD_LIBRARY_PATH setting for these unit tests?

And it seems like no debian build happens?

**Comment by manuelm** [comment]
@markus2330 As I said I'll take over the travis work as soon as the OS X build works. Don't - I repeat - don't merge any travis commits.

File: .travis.yml (hunk 1 above)
Comment (manuelm) [inline .travis.yml]: Because it's annoying while setting up travis..

File: .travis.yml (hunk 2 above)
Comment (manuelm) [inline .travis.yml]: No

File: scripts/docker/Dockerfile.debian_unstable (hunk 4 above)
Comment (manuelm) [inline scripts/docker/Dockerfile.debian_unstable]: This won't stay...

File: src/bindings/jna/CMakeLists.txt (hunk 5 above)
Comment (manuelm) [inline src/bindings/jna/CMakeLists.txt]: @mpranj Why aren't you fetching the jar in .travis.xml/before_install and just move it to /usr/share/java there?

**Comment by manuelm** [comment]
Yeah, looks like bindings are still broken.
And crypto_openssl is still excluded? Any chance to get this working?

**Comment by markus2330** [comment]
I think it is ok to exclude crypto_* for now: It has currently no functionality exposed to users and it is also tagged as experimental.

Btw. did you try to install elektra and exec `kdb run_all`? (Like in jenkins elektra-local-installation) Maybe we can this even do as root installed directly into system pathes as the container is recreated every time anyway? Then the result of the travis build would be a container where the latest Elektra is installed.

**Comment by manuelm** [comment]
Travis is unable to find the openssl dev headers. That's why I've excluded the plugin. Most probably this is just a missing path someone with a real OS X machine can fix quickly.

@markus2330 Can we please only discuss OSX related things here. And stop thinking about new things after everything (incl #730) is set up. There are multiple reasons why this is a bad idea.

**Comment by markus2330** [comment]
The installation and `kdb run` is related to our current problem (in Mac OS X): If we get the lua/python tests running more easily in the installed system, we could exclude them when running `make test` in the build directory.

On jenkins it was adding about two lines, `make install` and then `kdb run_all`. So it might be the easiest solution for this problem and additionally makes the tests more realistic. If it is not such a simple change, we should of course not add it in this PR.

**Comment by markus2330** [comment]
Btw. build tests should be after the installed tests because some build tests require an installed Elektra (external+context tests).

File: .travis.yml (hunk 2 above)
Comment (mpranj) [inline .travis.yml]: Sure we can build on all available versions on travis once we get it to work at least on the current version.

File: src/bindings/jna/CMakeLists.txt (hunk 5 above)
Comment (mpranj) [inline src/bindings/jna/CMakeLists.txt]: Isn't the jar bundled with elektra later such that the original jna.jar can be thrown away?
Otherwise yes, I can move it somewhere. I was trying not to magically move files into system directories just for `make` to work.

Comment (markus2330) [inline src/bindings/jna/CMakeLists.txt]: It is not bundled at the moment, but I think it makes sense because as far as I know the java world wants a jar to be self-contained?

Comment (mpranj) [inline src/bindings/jna/CMakeLists.txt]: I have no idea how they do it. The "clean" way I know is using maven but that doesn't seem to be a (clean) option here.

Comment (markus2330) [inline src/bindings/jna/CMakeLists.txt]: It is weird, https://cmake.org/cmake/help/v3.0/module/UseJava.html says that "For backwards compatibility, jar files listed as sources are ignored.", but it does not say anything about how to include them. INCLUDE_JARS only seems to set the include path, but not actually include them.

Actually we would only need to remove the line 451 of /usr/share/cmake-3.0/Modules/UseJava.cmake which does this strange ignoring of jar because of "backward compatibility".

File: src/plugins/crypto/CMakeLists.txt [hunk 7]
```diff
@@ -5,83 +5,33 @@ include (LibAddPlugin)
...
+endif(APPLE)
 
-		try_compile (HAS_OPENSSL_4SURE
-			"${CMAKE_BINARY_DIR}"
-			"${PROJECT_SOURCE_DIR}/src/plugins/crypto/compile_openssl.c"
-			CMAKE_FLAGS
-				-DINCLUDE_DIRECTORIES:STRING=${OPENSSL_INCLUDE_DIR}
-				-DLINK_LIBRARIES:PATH=${OPENSSL_LIBRARIES}
-		)
+find_package(OpenSSL)
```
Comment (mpranj) [inline src/plugins/crypto/CMakeLists.txt]: I had to find this outside of `DEPENDENCY_PHASE` otherwise `OPENSSL_FOUND` etc. get reset.

Comment (markus2330) [inline src/plugins/crypto/CMakeLists.txt]: It does not matter if the options are reset, the add_plugin remembers the arguments it got.

find_package should only be used in the DEPENDENCY_PHASE.

I think the OPENSSL_MANUAL_OSX_HACK already does what you are trying to achieve.

And as said: the crypto plugin has no priority for this PR, its better if you revert these changes.

Comment (mpranj) [inline src/plugins/crypto/CMakeLists.txt]: Sure.

**Comment by mpranj** [comment]
@markus2330 @manuelm just a quick update: `dlopen()` works now (FreeBSD/OS X). @markus2330 had almost done it in da243f9e25d8fa14f8286c48b4338a73c1e7242d but the path was wrong.

Also, I got it to build successfully etc. but it's not ready yet for @manuelm to take over, I will ping him later for that.

File: .travis.yml [hunk 8]
```diff
@@ -0,0 +1,69 @@
...
+  - mkdir build && cd build
+  - CMAKE_OPT=()
+  - echo $PATH
+  - if [[ "$TRAVIS_OS_NAME" == "osx" ]]; then
+      python2_ver=$(python2 -c 'import sys; print(".".join(map(str, sys.version_info[:2])))') &&
+      CMAKE_OPT+=("-DPYTHON2_INCLUDE_DIR:PATH=$(python2-config --prefix)/include/python${python2_ver}") &&
+      CMAKE_OPT+=("-DPYTHON2_LIBRARY:FILEPATH=$(python2-config --prefix)/lib/libpython${python2_ver}.dylib") &&
+      python3_ver=$(python3 -c 'import sys; print(".".join(map(str, sys.version_info[:2])))') &&
+      CMAKE_OPT+=("-DPYTHON_INCLUDE_DIR:PATH=$(python3-config --prefix)/include/python${python3_ver}m") &&
+      CMAKE_OPT+=("-DPYTHON_LIBRARY:FILEPATH=$(python3-config --prefix)/lib/libpython${python3_ver}.dylib");
```
Comment (mpranj) [inline .travis.yml]: While this does fix the builds on travis I would suggest/like to solve the same with cmake. It would make building on OS X a litte less painful.

The nice thing with travis is: we now have a raw description of the build dependencies which we could document in `doc/COMPILE.md` or somewhere.

Comment (manuelm) [inline .travis.yml]: I really don't like putting platform specific paths into *our* build system. If cmake is unable to find it, try to fix this at cmakes side or leave it. Otherwise you'll have to constantly keep up with platform specifics of other people, which fills our build system with obscure paths nobody can verify or clean up. Heck, right now cmake doesn't even find python3 on debian jessie or wheezy. And nobody cares about that, unless you're building this on our own. And even if you build elektra on your own, you usually pull the dependencies and configure line from the package files of your distribution and be done.

Beside that nobody stops a user from downloading python and installing into different paths. It just happens that brew is the easiest way on OSX so most users are using this packager right now. Might be another packager and other paths a few years later.

So instead, I suggest we do the same thing like on Linux: You (or a OSX maintainer) just write a definition file for brew in which you can put the special brew specific paths.

I have the same opinion regarding the automatic jna jar download.

Comment (mpranj) [inline .travis.yml]: I do not want to put OS X specific paths all over the build system.
I do however think that a python user will have python/python2/python3 somewhere in their PATH.

Using python-config works fine for me on FreeBSD and OS X:
```
mpranj@beastie ~ % python2-config --includes
-I/usr/local/include/python2.7 -I/usr/local/include/python2.7
mpranj@beastie ~ % python2-config --libs
-lpython2.7 -L/usr/local/lib -lintl -lutil -lm
mpranj@beastie ~ % python2-config --cflags
-I/usr/local/include/python2.7 -I/usr/local/include/python2.7 -fno-strict-aliasing -O2 -pipe -fstack-protector -fno-strict-aliasing -DNDEBUG
mpranj@beastie ~ % python2-config --ldflags
-L/usr/local/lib -lpython2.7 -L/usr/local/lib -lintl -lutil -lm -Wl,--export-dynamic
```

Comment (manuelm) [inline .travis.yml]: Than please fix it on cmakes side: https://cmake.org/Bug/bug_relationship_graph.php?bug_id=9897
However I do not know why cmake doesn't call this in the first place. Might be related to cross-compiling.

**Comment by mpranj** [comment]
I think this is failing only because of the deprecated github status updater, which I am currently removing from all jobs where it might be still in use.

This should be ready for @manuelm to implement the linux builds.

It solely concentrates on building for OS X on travis and introduces **no** fixes. For instance, I removed the automatic JNA download. It's a different matter. CMake doesn't even find JNA on FreeBSD where there is a nice port available. If you want us to take a look at this and make it work on more platforms, I will open a new issue.

**Comment by markus2330** [comment]
Yes, elektra-mergerequests-unstable seems to fail only because of the Github Status setter.

So should I merge it? It looks great: it seems like everything finally works on Mac OS X. Or are there also leftovers for Mac OS X?

Please create an issue for every known problem, it does not mean you have to work for it. When someone needs JNA on FreeBSD we can fix it.

**Comment by mpranj** [comment]
I removed the Github status setter. Let's see if it works well now.

No, @manuelm explicitly wanted us **not to merge**, but to take over the work and finish it. I made the PR just so it could be reviewed.

Other than that: yes I'd like to see this in master asap so we can detect regressions on OS X automatically.

**Comment by manuelm** [comment]
> So should I merge it? 

Once again: no!

@mpranj 
Will take at look next (or in two weeks).

I see not much sense in testing multiple xcode variants. Apple supports the latest only. And apart from different clang versions we don't use any SDK shipped with xcode. All it take is more time to build.

**Comment by mpranj** [comment]
@manuelm the different xcode variants are basically different VMs/different OS X versions. The three tested versions are still supported/get security updates and are in use. Not everyone has the latest OS X.

It makes sense because apple tends to change lots of things. One example: they introduced SIP https://en.wikipedia.org/wiki/System_Integrity_Protection which doesn't allow us to write to just any system directory. (also different library versions, OS X not using openssl any more, ..)
It might be interesting for the future if things get broken with OS X upgrades and it doesn't cost us anything right now. But you can remove it as you wish.

**Comment by markus2330** [comment]
I am also strongly in favour of testing all variants, it often gives insightful clues in which situation something breaks. We can always decide to drop support for some old system if it is hard to fix something (and it is obviously the systems fault), but I would not exclude it only to reduce build time.

It would be very useful to have Mac OS X builds now, and not to wait 1-2 weeks for some other feature which is not urgent. It is no problem if you completely rewrite the travis file later (if needed).

**Comment by manuelm** [comment]
Then for gods sake merge it. The .travis.xml only. Rest is not needed.

And yes, I'll definitely rewrite the whole damn thing.

**Comment by markus2330** [comment]
It does not really make sense to say someone should not merge but its okay to cherry-pick exactly the same commits?

I only added the travis file, as asked for.

We should not invest too much time in build-stuff but rather invest time into the code, if a buildjob works it is good enough.
